### DISCUSSION TITLE
Challenges with Extracting and Reapplying Specific Layers from Multi-layer PDFs using PyMuPDF

### DISCUSSION THREAD

**Discussion by Danishkhilji** [pymupdf/PyMuPDF#4091] [Looking for help]
Hello,

I’m working with a multi-layer PDF file, and I need to extract and reapply content from a specific layer (OCG) into a new PDF. Here's what I want to achieve:

1. When a user uploads a PDF file, I want to go through the pages and extract content from a specific layer (such as vectors, images, and text) based on the layer name.
2. I want to create a new PDF, where the extracted content from the specified layer is placed into a new optional content group (OCG) layer in the newly created file.
3. The new PDF should retain the same page structure and visual content, but only the data from the target layer should appear in the new OCG.

I have tried three different approaches to achieve this, but I am encountering issues in each of them. Here’s an overview of the approaches and the problems I’m facing:

**Approach 1: Copy Desired Objects and Paste into a New PDF with a Layer**
**Problem:**
I attempted to copy only the desired objects (vectors) from the specified layer and paste them into a new PDF with a new layer. However, I am unable to selectively copy vector objects. 

**Approach 2: Copy All Content and Remove Unwanted Objects, Then Paste into a Layer**
**Problem:**
In this approach, I tried copying all content from the original PDF using **show_pdf_page** and then removing unwanted objects by using methods like **add_redact_annot()** and **apply_redactions()**. While this approach successfully copied and pasted everything into a new file, it merged all the layer data into the root of the new PDF. When I tried removing unwanted layers, the method also deleted objects I wanted to keep, since some objects are present in overlapping areas across different layers. This caused the unwanted removal of content that was supposed to remain.

**Approach 3: Copy Drawings and Redraw on a New Layer**
**Problem:**
In the third approach, I copied the drawings (vectors) and redrew them on a new layer. While I successfully extracted and redrew the vector graphics from the target layer, this method also ended up redrawing areas that had been clipped out in the original file. I couldn’t figure out how to identify or preserve the clipping path, and thus, the redrawn content included areas that were supposed to be clipped. As a result, I ended up with content that didn’t match the original layout.

**Questions:**

1. How can I efficiently extract and reapply a specific layer’s content (such as vectors, images, and text) without unintentionally merging unwanted layers or areas?
2. Is there a way to handle clipped content (e.g., identify and preserve clipping paths) when redrawing vector objects onto a new layer?
3. Are there any better methods or recommendations for working with OCGs (Optional Content Groups) in PyMuPDF?

Thank you for any guidance or suggestions!

**Comment by JorjMcKie** [comment]
Well thought approaches!
And your results / conclusions are also correct: none of the alternatives will work in the general case.

I see no way to achieve what you want. Drawing with the `Shape` class supports no clipping as you also have determined - and this is not planned either.

The only thing that may come close to your intention is setting (combinations of) OCGs permanently active and then save the PDF in that state.

**Reply by Danishkhilji** [reply]
@JorjMcKie Thank you for the prompt response! I appreciate the acknowledgement of the approaches.

Regarding your suggestion:

> The only thing that may come close to your intention is setting (combinations of) OCGs permanently active and then saving the PDF in that state.


Could you please elaborate on this? Does it mean saving the PDF with the desired OCGs (layers) activated and then working with the saved file in that state for further processing? If so, would this allow for extracting content from just the active layers without affecting the rest of the document, or is the goal primarily to control visibility for visual purposes?

**Reply by JorjMcKie** [reply]
> Does it mean saving the PDF with the desired OCGs (layers) activated and then working with the saved file in that state for further processing? If so, would this allow for extracting content from just the active layers without affecting the rest of the document, or is the goal primarily to control visibility for visual purposes?

Exactly. This approach (= saving multiple copies of an original file, each copy with the desired layer combination activated) would yield not only file versions **_looking_** like desired, but also all extractions (text, images, graphics) would only select data which correspond to activated layers.

**Reply by Danishkhilji** [reply]
@JorjMcKie You're absolutely right! Using this approach—saving multiple copies of the original file with the desired layer combinations activated—does help in isolating content visually. It ensures that all extractions (text, images, and graphics) align with the activated layers, which is a great step forward.

I also tried a similar approach (**Approach 3**), and while it successfully copied the desired content, an issue arose during the redrawing process. Specifically, when the vectors were redrawn, the clipping paths were effectively disabled, causing previously hidden (clipped) portions of the vectors to become visible again.

**Comment by Danishkhilji** [comment]
I have tried following code snippet to extract drawing elements from a PDF page:

```
drawings = page.get_drawings(extended=True)
for i, draw in enumerate(drawings):
    if draw['type'] == 'clip' and draw['layer'] == 'Design':
        print(draw)
```


This code outputs objects with details like this:

```
{'items': [('l', Point(579.25, 301.31207275390625), Point(579.25, 375.31207275390625)), ('l', Point(579.25, 375.31207275390625), Point(604.7459716796875, 375.31207275390625)), ('c', Point(604.7459716796875, 375.31207275390625), Point(632.4329833984375, 359.154052734375), Point(653.4140014648438, 332.75701904296875), Point(662.4970092773438, 301.31207275390625)), ('l', Point(662.4970092773438, 301.31207275390625), Point(579.25, 301.31207275390625))], 'closePath': False, 'type': 'clip', 'even_odd': False, 'scissor': Rect(579.25, 301.31207275390625, 662.4970092773438, 375.31207275390625), 'level': 0, 'layer': 'Design'}
{'items': [('l', Point(511.7090148925781, 149.8370361328125), Point(511.7090148925781, 211.8370361328125)), ('l', Point(511.7090148925781, 211.8370361328125), Point(622.7100219726562, 211.8370361328125)), ('l', Point(622.7100219726562, 211.8370361328125), Point(622.7100219726562, 169.8270263671875)), ('c', Point(622.7100219726562, 169.8270263671875), Point(613.0079956054688, 161.6170654296875), Point(602.0419921875, 154.841064453125), Point(590.1580200195312, 149.8370361328125)), ('l', Point(590.1580200195312, 149.8370361328125), Point(511.7090148925781, 149.8370361328125))], 'closePath': False, 'type': 'clip', 'even_odd': False, 'scissor': Rect(511.7090148925781, 149.8370361328125, 622.7100219726562, 211.8370361328125), 'level': 0, 'layer': 'Design'}
{'items': [('re', Rect(1387.321044921875, 853.205078125, 1570.322021484375, 894.2040405273438), 1)], 'type': 'clip', 'even_odd': False, 'closePath': False, 'scissor': Rect(1387.321044921875, 853.205078125, 1570.322021484375, 894.2040405273438), 'level': 0, 'layer': 'Design'}
{'items': [('c', Point(277.8320007324219, 651.488037109375), Point(277.8320007324219, 738.632080078125), Point(348.4750061035156, 809.2760620117188), Point(435.6189880371094, 809.2760620117188)), ('l', Point(435.6189880371094, 809.2760620117188), Point(435.6189880371094, 493.7010498046875)), ('c', Point(435.6189880371094, 493.7010498046875), Point(348.4750061035156, 493.7010498046875), Point(277.8320007324219, 564.3440551757812), Point(277.8320007324219, 651.488037109375))], 'type': 'clip', 'even_odd': False, 'closePath': False, 'scissor': Rect(277.8320007324219, 493.7010498046875, 435.6189880371094, 809.2760620117188), 'level': 0, 'layer': 'Design'}
{'items': [('l', Point(680.5700073242188, 544.5890502929688), Point(680.5700073242188, 595.0570678710938)), ('c', Point(680.5700073242188, 595.0570678710938), Point(680.5700073242188, 600.248046875), Point(682.3829956054688, 604.655029296875), Point(686.0170288085938, 608.2870483398438)), ('c', Point(686.0170288085938, 608.2870483398438), Point(689.64697265625, 611.9190673828125), Point(694.0560302734375, 613.7340698242188), Point(699.2459716796875, 613.7340698242188)), ('c', Point(699.2459716796875, 613.7340698242188), Point(704.4329833984375, 613.7340698242188), Point(708.8410034179688, 611.9190673828125), Point(712.4739990234375, 608.2870483398438)), ('c', Point(712.4739990234375, 608.2870483398438), Point(716.10498046875, 604.655029296875), Point(717.9219970703125, 600.248046875), Point(717.9219970703125, 595.0570678710938)), ('l', Point(717.9219970703125, 595.0570678710938), Point(717.9219970703125, 581.9390258789062)), ('l', Point(717.9219970703125, 581.9390258789062), Point(755.7169799804688, 581.9390258789062)), ('l', Point(755.7169799804688, 581.9390258789062), Point(755.7169799804688, 1012.0100708007812)), ('c', Point(755.7169799804688, 1012.0100708007812), Point(755.7169799804688, 1017.2000732421875), 
Point(757.5689697265625, 1021.6090698242188), Point(761.2760009765625, 1025.239013671875)), ('c', Point(761.2760009765625, 1025.239013671875), Point(764.97998046875, 1028.8720703125), Point(769.5020141601562, 1030.68701171875), Point(774.8380126953125, 1030.68701171875)), ('c', Point(774.8380126953125, 1030.68701171875), Point(779.8759765625, 1030.68701171875), Point(784.25, 1028.8720703125), Point(787.9550170898438, 1025.239013671875)), ('c', 
Point(787.9550170898438, 1025.239013671875), Point(791.6599731445312, 1021.6090698242188), Point(793.5139770507812, 1017.2000732421875), Point(793.5139770507812, 1012.0100708007812)), ('l', Point(793.5139770507812, 1012.0100708007812), Point(793.5139770507812, 581.9390258789062)), ('l', Point(793.5139770507812, 581.9390258789062), Point(831.3099975585938, 581.9390258789062)), ('l', Point(831.3099975585938, 581.9390258789062), Point(831.3099975585938, 595.5020751953125)), ('c', Point(831.3099975585938, 595.5020751953125), Point(831.3099975585938, 600.5430297851562), Point(833.2000122070312, 604.9160766601562), Point(836.97998046875, 608.6190185546875)), ('c', Point(836.97998046875, 608.6190185546875), Point(840.7589721679688, 612.3260498046875), Point(845.2410278320312, 614.1770629882812), Point(850.4320068359375, 614.1770629882812)), ('c', Point(850.4320068359375, 614.1770629882812), Point(855.468994140625, 614.1770629882812), Point(859.843017578125, 612.3650512695312), Point(863.5479736328125, 608.7310791015625)), ('c', Point(863.5479736328125, 608.7310791015625), Point(867.2520141601562, 605.10107421875), Point(869.1069946289062, 600.6920776367188), Point(869.1069946289062, 595.5020751953125)), ('l', Point(869.1069946289062, 595.5020751953125), Point(869.1069946289062, 544.5890502929688)), ('l', Point(869.1069946289062, 
544.5890502929688), Point(680.5700073242188, 544.5890502929688))], 'closePath': False, 'type': 'clip', 'even_odd': False, 'scissor': Rect(680.5700073242188, 544.5890502929688, 869.1069946289062, 1030.68701171875), 'level': 0, 'layer': 'Design'}
```


These entries have the type clip, and their structure includes paths (e.g., l for line, c for cubic Bézier curves), scissor rectangles, and other attributes. My goal is to determine if these are clipping paths.

1. Are these objects actual clipping paths in the context of the PDF?
2. If yes, I'm struggling to figure out how to apply these clipping paths when redrawing the page using PyMuPDF. I reviewed the Shape methods in PyMuPDF, but I couldn't find a clear way to apply these clipping paths.

Any guidance or examples on how to apply these paths during a redraw process would be greatly appreciated.

**Reply by JorjMcKie** [reply]
Clipping is not supported at all by the `.draw_*` methods - sorry.
